cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsCloudWatchService.java

```diff
@@ -74,35 +75,51 @@ public void addCloudWatchAlarmsForSystemFailures(List<CloudResource> instances,
                 .collect(Collectors.toList());
+        LOGGER.info("Found ids for running instances: [{}]. Cloudwatch alarms for these instances will be deleted.", instanceIds);
+        List<String> deletedInstanceIds = stack.getGroups().stream()
+            .flatMap(group -> group.getDeletedInstances().stream())
+            .map(CloudInstance::getInstanceId)
+            .collect(Collectors.toList());
+        LOGGER.info("Found ids for deleted instances: [{}]. Deletion will be attempted on the cloudwatch alarms for these instances.", deletedInstanceIds);
+        instanceIds.addAll(deletedInstanceIds);
         deleteCloudWatchAlarmsForSystemFailures(stack, regionName, credentialView, instanceIds);
     }
 
     public void deleteCloudWatchAlarmsForSystemFailures(CloudStack stack, String regionName, AwsCredentialView credentialView, List<String> instanceIds) {
-        List<String> instanceIdsFromStack = stack.getGroups().stream()
+        if (instanceIds == null || instanceIds.isEmpty()) {
+            LOGGER.warn("No instance ids provided for cloudwatch alarm deletion. No alarms will be deleted.");
+        } else {
+            LOGGER.info("Deleting alarms for instance ids [{}]", instanceIds);
+            List<String> instanceIdsFromStack = stack.getGroups().stream()
                 .flatMap(group -> group.getInstances().stream())
                 .map(CloudInstance::getInstanceId)
                 .collect(Collectors.toList());
-        List<String> instanceIdsNotInStack = instanceIds.stream()
+            List<String> instanceIdsNotInStack = instanceIds.stream()
                 .filter(instanceId -> !instanceIdsFromStack.contains(instanceId))
                 .collect(Collectors.toList());
-        if (!instanceIdsNotInStack.isEmpty()) {
-            LOGGER.warn("Instance IDs [{}] are not part of cloud stack {}, these instances may have already been deleted on the cloud provider side.",
-                    instanceIdsFromStack, stack);
+            if (!instanceIdsNotInStack.isEmpty()) {
+                LOGGER.warn("Instance IDs [{}] are not part of cloud stack {}, these instances may have already been deleted on the cloud provider side.",
+                    instanceIdsNotInStack, stack);
+            }
+            deleteCloudWatchAlarmsForSystemFailures(regionName, credentialView, instanceIds);
         }
-        deleteCloudWatchAlarmsForSystemFailures(regionName, credentialView, instanceIds);
     }
 
     private void deleteCloudWatchAlarmsForSystemFailures(String regionName, AwsCredentialView credentialView, List<String> instanceIds) {
+        // The list of alarms received may be longer than 100 items, but alarms can only be deleted in batches of max
+        // size 100. To work around this, break the instanceIds list into chunks no greater than 100, and process each
+        // chunk. See https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_DeleteAlarms.html
```

Comment:
**praise:** Thanks for explaining this batch size, which would be pretty obtuse otherwise!